diff -crB curl-7.20.1/include/curl/curl.h curl-7.20.1-patched/include/curl/curl.h
*** curl-7.20.1/include/curl/curl.h	2010-03-24 08:41:07.000000000 -0400
--- curl-7.20.1-patched/include/curl/curl.h	2010-06-14 19:56:42.000000000 -0400
***************
*** 167,172 ****
--- 167,173 ----
  };
  
  typedef int (*curl_progress_callback)(void *clientp,
+                                       int connected,
                                        double dltotal,
                                        double dlnow,
                                        double ultotal,
diff -crB curl-7.20.1/lib/progress.c curl-7.20.1-patched/lib/progress.c
*** curl-7.20.1/lib/progress.c	2010-03-24 08:41:08.000000000 -0400
--- curl-7.20.1-patched/lib/progress.c	2010-06-14 19:57:41.000000000 -0400
***************
*** 340,345 ****
--- 340,346 ----
        /* There's a callback set, so we call that instead of writing
           anything ourselves. This really is the way to go. */
        result= data->set.fprogress(data->set.progress_client,
+                                   data->progress.flags ? 1 : 0,
                                    (double)data->progress.size_dl,
                                    (double)data->progress.downloaded,
                                    (double)data->progress.size_ul,
